app/controllers/api/v1/user_annotations_controller.rb

```diff
@@ -0,0 +1,119 @@
+          key :description, 'Subsample threshold'
+        end
+        property :subsample_annotation do
+          key :type, :string
+          key :description, 'Subsample annotation'
+        end
+      end
+
+      swagger_path '/studies/{accession}/user_annotation' do
+        operation :post do
+          key :tags, [
+              'Studies'
+          ]
+          key :summary, 'Create user annotation'
+          key :description, 'Create new custom user annotation for the study'
+          key :operationId, 'site_study_user_create_annotation_path'
+          parameter do
+            key :name, :accession
+            key :in, :path
+            key :description, 'Study accession number (e.g. SCP123)'
+            key :required, true
+            key :type, :string
+          end
+          parameter do
+            key :name, :user_annotation
+            key :in, :body
+            key :description, 'UserAnnotationInput object'
+            schema do
+              key :'$ref', :UserAnnotationInput
+            end
+          end
+          response 200 do
+            key :description, 'Successfully created user annotation'
+          end
+          extend SwaggerResponses::StudyControllerResponses
+        end
+      end
+
+      def create
+
+        begin
+          # Parameters to log for any errors
+          user_annotation_params = params.dup
+          user_annotation_params[:current_user] = current_user
+          log_params = user_annotation_params.dup
+          # Don't log data_arrays; it's too big
+          log_params.delete(:user_data_arrays_attributes)
+
+          cluster = params[:cluster]
+
+          message, annotations = UserAnnotationService.create_user_annotation(
+            @study, params[:name], params[:user_data_arrays_attributes],
+            cluster, params[:loaded_annotation],
+            params[:subsample_threshold], current_api_user
+          )
+
+          render json: {message: message, annotations: annotations}
+
+        # Handle other errors in saving user annotation
+        rescue Mongoid::Errors::InvalidValue => e
+          error_context = ErrorTracker.format_extra_context(@study, {params: log_params})
+          ErrorTracker.report_exception(e, current_user, error_context)
+          message = 'The following errors prevented the annotation from being saved: ' + 'Invalid data type submitted. (' + e.problem + '. ' + e.resolution + ')'
+          Rails.logger.error "Creating user annotation of params: #{user_annotation_params}, invalid value of #{e.message}"
+          render json: {error: error}, status: 400 # Bad request
+
+        rescue => e
+          # If a generic unexpected error occurred and couldn't save the annotation
+          error_context = ErrorTracker.format_extra_context(@study, {params: log_params})
+          ErrorTracker.report_exception(e, current_user, error_context)
+
+          if e.summary.include? 'Name' and e.summary.include? 'has already been taken'
```

Comment:
This likewise feels like specific business logic that belongs in the service layer -- the service layer can catch this, and throw it as an ArgumentError.  Then this controller can do `rescue ArgumentError { tell the user it's their own fault }` and `rescue  <<Everything else>> { tell the user it was something unexpected }`

I hope I'm being clear about the general pattern I want -- service methods should handle the business logic, and re-throw low-level exceptions as the correct type of exception that would make sense to the caller.  E.g. if a caller passes in bad values to a service method, the service should throw ArgumentError, because that's the exception that makes sense to the caller, regardless of what exception caused the service method to realize that one of its arguments was bad.  

Happy to hop on a call to chat--this is something that's relatively important as we continue to convert to this pattern of thin controllers delegating to business services, and so I want to make sure we're on the same page.

Reply:
Good points, and thanks for the chat -- done in 655b981.